### DISCUSSION TITLE
Proper way to handle concave kinetic meshes?

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#1001] [Q&A] [answered]
I'm converting a free mobile game of mine from Bullet to Jolt. This video probably demonstrates it better than I could ever explain here. In Bullet, this was a GImpact TriangleMeshShape.
Is the way-to-go for this kind of a shape to deconstruct it with V-HACD and then use compound convex shapes?

https://github.com/jrouwe/JoltPhysics/assets/159929665/1572256b-eaf7-44ac-8fe7-6851376bb566

**Comment by jrouwe** [comment]
The only restriction currently is that MeshShapes cannot collide with other MeshShapes or HeightfieldShapes. So as long as your character and the other dynamic objects in this level are convex then the wall can be a regular kinematic MeshShape. Decomposing using V-HACD is also possible and could result in a better simulation when your dynamic objects are moving fast (because a convex mesh has a better concept of inside and outside so is more likely to resolve the collision correctly if the object is embedded in the wall)

**Reply by SSNTails** [reply]
When I create kinematic concave shapes, I get an assert failure about mass being zero:
https://github.com/jrouwe/JoltPhysics/blob/e8a99bfa0c60997eb4a7c22bd5c34f36a87dc632/Jolt/Physics/Body/MotionProperties.cpp#L29

Is this a bug (I thought kinematic shapes had infinite mass? Or is it correct that you manually need to supply some kind of mass via the 
BodyCreationSettings.mMassPropertiesOverride ?

`
>	JPH::MotionProperties::SetMassProperties(JPH::EAllowedDOFs inAllowedDOFs, const JPH::MassProperties & inMassProperties) Line 29	C++
 	JPH::BodyManager::AllocateBody(const JPH::BodyCreationSettings & inBodyCreationSettings) Line 221	C++
 	JPH::BodyInterface::CreateBody(const JPH::BodyCreationSettings & inSettings) Line 22	C++
 	JPH::BodyInterface::CreateAndAddBody(const JPH::BodyCreationSettings & inSettings, JPH::EActivation inActivationMode) Line 154	C++
 	CollisionHandler::AddTriangleMesh(WObj::WObj * object) Line 618	C++
 	WObj::DoObjectSpawn(void * pSpawnobj) Line 432	C++
 	SpawnObj::Spawn() Line 66	C++
 	SpawnSpawnobjs() Line 80	C++
 	World::Load(const char * levelPath) Line 392	C++
 	ChangeLevel(int argc, char * * argv) Line 54	C++
 	SV::ExecuteCommand(char * cmd) Line 236	C++
 	SV::ExecuteQueue() Line 265	C++
 	Update(float deltaTime) Line 302	C++
 	WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 2132	C++
 	[External Code]	
`


Creation code, just for completeness:
`

	JPH::BodyCreationSettings bcs(new JPH::MeshShapeSettings(triangles), mat.GetTranslation(), mat.GetRotation().GetQuaternion(), JPH::EMotionType::Kinematic, Layers::MOVING);
	bcs.mUserData = (JPH::uint64)object;
	bcs.mRestitution = 0;
	bcs.mAngularDamping = 0;
	bcs.mFriction = 0;
	bcs.mMaxAngularVelocity = 0;
	JPH::MassProperties msp;
	msp.ScaleToMass(10); // arbitrary???

	bcs.mMassPropertiesOverride = msp;
	bcs.mOverrideMassProperties = JPH::EOverrideMassProperties::CalculateInertia;

	object->collisionObjectID = bi.CreateAndAddBody(bcs, JPH::EActivation::Activate);
`

**Reply by jrouwe** [reply, marked as answer]
ScaleToMass only works if it has mass to begin with (X * 0 == 0).

Do something like this (the actual mass doesn't matter for a kinematic object as long as it is not zero):

https://github.com/jrouwe/JoltPhysics/blob/0191c72d5f6a66217df02c0f6ad5d1c63134b4d4/Samples/Tests/General/DynamicMeshTest.cpp#L32-L33

**Reply by SSNTails** [reply]
Great, thanks. The code I posted actually DOES work, though. It's when I exclude the mMassPropertiesOverride that I end up with the assert failure of 0.

Just as an API suggestion, if a kinematic body is added with a mass of zero, should the API automatically handle that, since mass doesn't matter for a kinematic object anyway?

**Reply by jrouwe** [reply]
> Just as an API suggestion, if a kinematic body is added with a mass of zero, should the API automatically handle that, since mass doesn't matter for a kinematic object anyway?

The problem is that at that point I don't know if the body is going to be switched to dynamic later, so ignoring it here may cause a problem later.
